src/metadata/firmware_download.rs

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+// SPDX-FileCopyrightText: 2025 1BitSquared <info@1bitsquared.com>
+// SPDX-FileContributor: Written by Rachel Mant <git@dragonmux.network>
+// SPDX-FileContributor: Modified by P-Storm <pauldeman@gmail.com>
+
+use std::ffi::OsStr;
+use std::path::PathBuf;
+
+use color_eyre::Result;
+use color_eyre::eyre::eyre;
+use serde::Deserialize;
+use url::Url;
+
+#[derive(Deserialize, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct FirmwareDownload
+{
+	#[serde(rename = "friendlyName")]
+	pub friendly_name: String,
+	#[serde(rename = "fileName")]
+	pub file_name: PathBuf,
+	pub uri: Url,
+}
+
+impl FirmwareDownload
+{
+	pub fn build_documentation_url(&self) -> Result<Url>
+	{
+		// Convert the path compoment of the download URI to a Path
+		let mut docs_path = PathBuf::from(&self.uri.path());
+		if docs_path.extension() == Some(OsStr::new("elf")) {
+			docs_path.set_extension("md");
+		} else {
+			return Err(eyre!("Suspected"));
```

Comment:
What is suspected? Overall this looks good but could do with a better error message here as this string is directly printed on the console when propagated.